synapseutils/sync.py

```diff
@@ -589,188 +594,138 @@ def _order_items(items):
+                        self._upload_item_async(
+                            upload_item.entity,
+                            upload_item.used,
+                            upload_item.executed,
+                            upload_item.store_kwargs,
+                            finished_items,
+                        )
                     )
-
-                    if used_pending or executed_pending:
-                        # we can't upload this item yet, it has provenance that
-                        # hasn't yet been uploaded
-                        skipped_items.append(item)
-                        pending_provenance.update(used_pending.union(executed_pending))
-
-                        # skip uploading because dependent provenance hasn't
-                        # finished uploading
-                        continue
-
-                # else not continued above due to pending provenance
-                # all provenance that this item depends on has already been uploaded
-                # so we can go ahead and upload this item
-
-                # we acquire the semaphore to ensure that we aren't uploading more
-                # than our configured maximum number of files here at once. once we
-                # reach the limit we'll block here until one of the existing file uploads completes
-                self._file_semaphore.acquire()
-                trace.get_current_span().set_attributes(
-                    {"thread.id": threading.get_ident()}
                 )
-                future = self._executor.submit(
-                    self._upload_item,
-                    item,
-                    used,
-                    executed,
-                    finished_items,
-                    pending_provenance,
-                    dependency_condition,
-                    abort_event,
-                    progress,
-                    context.get_current(),
-                )
-                futures.append(future)
-
-            with dependency_condition:
-                if pending_provenance.has_pending():
-                    # skipped_items contains all the items that we couldn't upload
-                    # the previous time through the loop because they depended on
-                    # another item for provenance. wait until there at least one
-                    # those items finishes before continuing another time through the loop.
-                    if not abort_event.is_set():
-                        dependency_condition.wait_for(
-                            lambda: (
-                                pending_provenance.has_finished_provenance()
-                                or abort_event.is_set()
-                            )
-                        )
-
-                pending_provenance.reset_count()
-
-            ordered_items = skipped_items
+                initial_items_to_pop.append(key)
+        for key in initial_items_to_pop:
+            ordered_items.pop(key)
```

Comment:
Why not just remove the dictionary items directly at the end of the `if not value:` block?

Reply:
Updated in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1083/commits/699c015fd8eef415e13b95a39e434f1ef8d6a323